test: create random and coins utils, add amount helper, dedupe add_coin

jonatack opened the pull request:
- Move random test utilities from `setup_common` to a new `random` file, as many tests don't use this code.

- Create a helper to generate semi-random CAmounts up to `MONEY_RANGE` rather than only uint32, and use the helper in the unit tests.

- De-duplicate a shared `add_coin` method by extracting it to a `coins` test utility.

jonatack committed:
Create InsecureRandMoneyAmount() test util helper

to generate semi-random CAmounts up to MAX_MONEY rather
than only uint32, and use it in the unit tests.

john-moffett commented on src/test/util/setup_common.cpp:
```diff
@@ -91,6 +93,16 @@ void Seed(FastRandomContext& ctx)
+    case 0: return 0;
+    case 1: return 1;
+    case 2: return MAX_MONEY;
+    default: return InsecureRandRange(InsecureRandBool() ? COIN : MAX_MONEY);
```
Can you explain the motivation for restricting the range to `COIN` half the time? Is it just because they're more common values?

jonatack replied:
Thanks for looking! Yes, the idea was to often generate a real-life value, while still testing the full `MoneyRange` in `src/consensus/amount.h`.

jonatack replied:
Updated its Doxygen doc with that info.

john-moffett reviewed (approved):
utACK 3cd46094d159373bda7efbaaf8c28bb78397947f

I like the idea of having more intelligent "random" data that includes edge cases a significant portion of the time.

fanquake commented on src/test/util/setup_common.cpp:
```diff
@@ -91,6 +93,16 @@ void Seed(FastRandomContext& ctx)
     ctx = FastRandomContext(seed);
 }
 
+CAmount SemiRandMoneyAmount()
```
No thoughts on the change, but not sure this code should be in `setup_common`. Probably `src/test/util/something` ?

jonatack replied:
Done. Didn't see a good existing home for it, so I moved the random generation test code from `setup_common` to `src/test/util/random`, which seems beneficial as many of the unit tests don't call it, and then added the test helper there.

jonatack committed:
Move random test util code from setup_common to random

as many of the unit tests don't use this code

jonatack commented:
Updated with feedback from @john-moffett and @fanquake (thanks!)

fanquake commented on src/Makefile.test_util.include:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2013-2019 The Bitcoin Core developers
+# Copyright (c) 2013-2023 The Bitcoin Core developers
```
There's no need to manually bump dates, and they'll likely just be removed (#26817), so you can drop any changes like this.

jonatack replied:
Done.

MarcoFalke commented on src/test/util/random.h:
```diff
@@ -0,0 +1,54 @@
+ * Generate a semi-random CAmount up to COIN (a typical value) or MAX_MONEY (the
+ * full money range), with occasional edge case values of 0, 1, and MAX_MONEY.
+ */
+static inline CAmount SemiRandMoneyAmount()
```
Not sure if we should imitate a fuzz engine in the unit tests to  pick special values. A fuzz engine is generally a lot better at finding those.

So I'd say, if you see value in testing more edge cases, I'd recommend to write a fuzz test, not a unit test with a hand written "edge-case finder".

jonatack replied:
I thought about that, but unlike probing for a crash as fuzzers do, our unit tests are using the rand generators to check returned values, i.e. non-crash behavior, from those inputs and this change follows and improves on their intention.

We could simplify the helper to just `InsecureRandRange(MAX_MONEY)` if reviewers prefer, but I think the proposed version provides more robust and useful coverage.

MarcoFalke replied:
> unlike probing for a crash as fuzzers do, our unit tests are using the rand generators to check returned values

Conceptually unit tests and fuzz tests are no different, if you take away the fuzz engine. A fuzz test doing a check is no different than a unit test doing a check. There may be an implementation detail in the exit code (and other stuff) if one calls assert vs BOOST_REQUIRE or BOOST_CHECK, but in both cases the test exe will end unsuccessfully.

> I think the proposed version provides more robust and useful coverage.

Would be good to provide an example where this leads to increased line or branch coverage

jonatack replied:
> We could simplify the helper to just `InsecureRandRange(MAX_MONEY)` if reviewers prefer

Happy to update with a concrete suggestion.

fanquake replied:
> Would be good to provide an example where this leads to increased line or branch coverage

I think this is the key point. If we're going to write special randomness/edge case functions, we should be able to show they are improving test coverage. Otherwise have a fuzzer pick random values.

jonatack replied:
Removed the edge cases. It uses the money range rather than only uint32, while remaining in the same vein as the existing unit test rand helpers.

maflcko reviewed (commented):
No objection, but not seeing the benefit over a fuzz test either?

jonatack committed:
De-duplicate add_coin methods to a test util helper

jonatack commented:
Rebased, squashed the money helper changes to one commit, and de-duplicated a shared `add_coin` method by extracting it to a test utility.

pinheadmz commented on src/test/util/blockfilter.cpp:
```diff
@@ -28,4 +28,3 @@ bool ComputeFilter(BlockFilterType filter_type, const CBlockIndex* block_index,
     filter = BlockFilter(filter_type, block, block_undo);
     return true;
 }
-
```
Why remove the trailing newline?

jonatack replied:
We don't normally have trailing newlines in files.

pinheadmz commented:
concept ACK
tested and reviewed, I like the deliberate introduction of edge cases into the tests and certainly makes sense to refactor out `AddTestCoin()` since it is duplicated. I'm not entirely sure about the move of all the random functions is a net benefit, you had to include that in 33 files when those functions were already sitting nicely in `setup_common`

pinheadmz commented:
ACK d34bc34c8c69781c913faa3e572af8030d099551

The location of the random functions isn't a blocker for me.

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK d34bc34c8c69781c913faa3e572af8030d099551
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmPhNmYACgkQ5+KYS2KJ
yTqGYA//cUrRxVOTho65ZbiTMKbev32/NMPI4aTt+j4Jl3UtSt96kqlBgSYdGftC
1V4B6pj2efxjFEuGQd8n1pzUt5pm8bkdg1r6JfVjEFNaz/ZqzagJS/93RU4/+3Ui
ffYt1jLI55ws2iy/owStLNmWqyRmh2YazTRPb+h1YAQ8FsuyIlpz75HJnKORsoA4
gcZwJ84YxSPaYYyCLpQ/++0PcofcR7q2CFidnZ8Dlcc3LwNSbdEk7zOwSkdAbT51
KIjIWH8Zh0LAgMbWCyXnmf1er3IdnCW/DSczQRqqbr+Mpn+0iQvyKsfqS+wJMOFj
Yjjg7aZ8NBx4nCw9mK+Ey3/XO1oAqcNVjRJinTKW7I6WhxShBY2vr9sF9ttX4nSS
7mEQf6iYRkCPCkRDrgTCtdNMzXtVQQRwZm+vN9w5Rs7AHz7SBJFDCpxC2hnMa1ij
FkowcK/FW1/LYYTceJmvwwbP5UuM5JN6z8j4DE6m/wztkiAOM0JSBe6IushlPplT
xeByjBkj3HEUoRStd4bZhZindIDplqHnzBEbqMuJFY8rLXrN2Vr6pP4esuUHlmh4
Nnm7In2py8RABm5sjeMugb5ElZR+zyDwnUElfqhwQf7iVKTqBfV+Jw05dH0Xmh9l
dM/qOScbDnDlg1IaIkHvFTkJSTirdb5pwalG5dg1i+rjsYeDoxw=
=2W5E
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

jonatack commented:
Trivial rebase per `git range-diff 52ddbd5 d34bc34 e6de5ad` for the merge of #26345.  Thank you for the ACKs, @john-moffett and @pinheadmz, mind re-ACKing?

I don't have a strong opinion on extracting the random utils to a helper.  As most of the test files don't need them to be a part of their compilation unit, it might be beneficial to speed up builds a bit, and it provides clearer code organization. OTOH it's more to review so am happy also to leave them in `setup_common` or extract only the new helper.

pinheadmz commented:
ACK e6de5ad8dce4fea053f75de5ee0189ed5f91df41

verified rebase change is minimal, also double checked that all `add_coin()` calls were updated in the unit tests

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK e6de5ad8dce4fea053f75de5ee0189ed5f91df41
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmPic+UACgkQ5+KYS2KJ
yToqeA//Uf61CfE9JSLcqfWT2IOu2J7AsFKY09zsXG+TpQQz1NiVOUvKPkQC0RYA
MdPnvrXTodlsYxf1c1K9xbAkmeQ356ptlcl/sDi5w5y2eQ82Ul5kFIrfEpPMB7M8
qfkcIdRsz3Ctv9ps0wbfeXSeW0rw2LbokQ8qhChtyzJAeUbyTDkjvjHq9vEMNs8G
QiGRFDEZzwxj+VkmsKOhYfJvOKUXXZJDGTzbUaATuLdXkbB17pE06n4nOAGUqbJa
rFbSC8EZVdwBp4Ryh3jtF7+FXlDizjh1Wc+Mushx1/LvBA2vRkz4U39e97VIRTEW
wHPvLkSyCtUX6HxuOCC2hGQY0mMfrxrtMTgUUQvTwIdu8S7JEHeEmQ2GhAGJ1T67
Bsnih3FMpHN7VI6ugek1Egdu08CXO5ybU6hyJNgaRTsIC29XlEmYVBGqj5XuyHsO
TYLStNA9A2cgHr/HfTrR/FB7RwaFqA47uPBvOapwmdV0dK7a1tCfJHrQxTO+SNkF
Gdg/gBE+yYqjEgA9uqa9t6rHPrQnFtdjG3Bp0AVFJ8WSh3l8P/ZJjmtGvYEFd7xt
LP6EEGPnt/vfmXnmViqDSmASRqnIa7sDdWcI92CG9Ds3+77SeHpFzRfPBH0zN3bi
X/zMr8XjCxbw8hswNUxSve4yc0KW5Nu2jTiv9zfJTSfoLYdDhC8=
=3Nhu
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

jonatack commented:
> If a majority of the values are now `0` and `1`

Only 1% was 0 and 1% was 1. I liked that it would check edge cases often enough to provide feedback when running `make check` that our current tests may miss, especially for developers running the unit tests locally to verify they are green, say, before pushing an update.

Anyway, that's gone now; removed the edge cases in the last push. It uses the money range rather than only uint32, while remaining in the same vein as the existing unit test rand helpers, and updated the PR description.

john-moffett commented on src/test/util/random.h:
```diff
@@ -0,0 +1,45 @@
+    return g_insecure_rand_ctx.randbool();
+}
+
+static inline CAmount SemiRandMoneyAmount()
```
If this is now just returning an amount in the uniform `[0 .. MAX_MONEY)` range, maybe rename to `InsecureRandMoneyAmount()`? Also, bit of a nit, but maybe have it return `InsecureRandRange(MAX_MONEY + CAmount(1))`, since `MAX_MONEY` is a valid `CAmount`?

john-moffett reviewed (approved):
ACK 4275195606e6f42466d9a8ef766b3035833df4d5

pinheadmz commented:
ACK 4275195606e6f42466d9a8ef766b3035833df4d5
<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 4275195606e6f42466d9a8ef766b3035833df4d5
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmPmX4QACgkQ5+KYS2KJ
yTp9Qg/8DS+6sPz95i2whVHQhnphYYAbbF68udZnle8PqeKof7lWGHOnCF3our+/
m2oOGfwXdviYMtIJzB+qBpwRHWonJStIl90Tz6yWqwnM2H5djWhpRBKUsjXMSPsI
vpFJCL2eQNSohNcBJ3aG2rZ3SIrMRZaFBiC1cKKan5XYbZBistXFg5H7odmTTwa/
j70qkh3dK+NINELUC5Q+exVSPK9LSwH8tmfIKjd0mb0RL7ABXISEVEQSjkdIgOG9
tNFc9o3Nlcseyha0V7bzkDEL0cSPoxInbxlUHv/o8+6+XnPL3cn2hGrFy424OMg6
IvZeiQacQbaHzDjDNI9bcK4nyunLqPz1G3r53ibs2CU507Mfv/TNIoToFyJzg/HF
kgb9qRhg/KwqaZ/2bT4OatAOja+t8lnFMvVL0NB0Uhmudv1kbcGERqIQWc8/0dru
17vdbiu3YqAstDMhtndvy/hupEuBEqEc97xNmXu9xmB8TP9FKDpRS2qWWpvbQNna
7gh49R+eBu/0kxeMhIrc2oXL1S3oaVH1kNqYwGT6cM7Qvt1nVJgGohm2JCE9PuPf
Md+VcMm/EhQwUjl7Ja58Sv3UdkWMmQKFj/DhVowJHht4tfDPtl6hR6iBitnB3/69
aoh/KBaiey/cSOXVv7zX5TQv21452+fdDuZPjFeGPct+vHxf/II=
=r6pu
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

achow101 commented:
ACK 4275195606e6f42466d9a8ef766b3035833df4d5

glozow commented:
What was the rationale for 81f5ade2a324167c03c5ce765a26bd42ed652723? test/util/random.h depends on test/util/setup_common because it's using `g_insecure_rand_ctx`, which means setup_common can't use any of the random functions without creating a circular dependency. Was this intentional?
